src/rgw/rgw_rest_s3.h

```diff
@@ -658,4 +664,120 @@ static inline int valid_s3_bucket_name(const string& name, bool relaxed=false)
   return 0;
 }
 
-#endif /* CEPH_RGW_REST_S3_H */
+class RGWS3V2TokenBasedAuthEngine : public RGWTokenBasedAuthEngine {
+  void get_other_tokens();
+
+protected:
+  req_state * const s;
+  std::string auth_sign;
+  std::string date;
+  bool qsr;
+  time_t now;
+
+  RGWS3V2TokenBasedAuthEngine(req_state * const s,
+                              CephContext * const cct,
+                              const Extractor& extr)
+    : RGWTokenBasedAuthEngine(cct, extr),
+      s(s),
+      qsr(false) {
+    get_other_tokens();
+    time(&now);
+  }
+
+  bool is_applicable() const noexcept override;
+};
+
+class RGWS3V2TokenExtractor : public RGWS3V2TokenBasedAuthEngine::Extractor {
+protected:
+  const req_state * const s;
+
+public:
+  RGWS3V2TokenExtractor(const req_state * const s)
+    : s(s) {
+  }
+  std::string get_token() const override;
+};
+
+class RGWS3V2LocalAuthEngine: RGWS3V2TokenBasedAuthEngine
+{
+protected:
+  RGWRados* const store;
+  const RGWLocalAuthApplier::Factory * const apl_factory;
+
+public:
+  RGWS3V2LocalAuthEngine(req_state * const s,
+                          RGWRados* const store,
+                          Extractor &ex,
+                          const RGWLocalAuthApplier::Factory * const apl_factory)
+    : RGWS3V2TokenBasedAuthEngine(s, s->cct, ex),
+      store(store),
+      apl_factory(apl_factory) {
+  }
+
+  const char* get_name() const noexcept override {
+    return "RGWS3V2LocalAuthEngine";
+  }
+  bool is_applicable() const noexcept override;
+  RGWAuthApplier::aplptr_t authenticate() const override;
+};
+
+class S3AuthFactory : public RGWRemoteAuthApplier::Factory,
+                      public RGWLocalAuthApplier::Factory {
+  typedef RGWAuthApplier::aplptr_t aplptr_t;
+  RGWRados * const store;
+  const std::string acct_override;
+
+public:
+  S3AuthFactory(RGWRados * const store,
+                const std::string& acct_override)
+    : store(store),
+      acct_override(acct_override) {
+  }
+
+  aplptr_t create_apl_local(CephContext * const cct,
+                            const RGWUserInfo& user_info,
+                            const std::string& subuser) const override {
+      return aplptr_t(
+        new RGWThirdPartyAccountAuthApplier<RGWLocalAuthApplier>(
+          RGWLocalAuthApplier(cct, user_info, subuser),
+          store, acct_override));
+    }
+  aplptr_t create_apl_remote(CephContext * const cct,
+                             RGWRemoteAuthApplier::acl_strategy_t&& acl_alg,
+                             const RGWRemoteAuthApplier::AuthInfo info
+                            ) const override {
+    return aplptr_t(
+        new RGWThirdPartyAccountAuthApplier<RGWRemoteAuthApplier>(
+        RGWRemoteAuthApplier(cct, store, std::move(acl_alg), info),
+        store, acct_override));
+  }
+};
+
+class RGWS3KeystoneAuthEngine: RGWS3V2TokenBasedAuthEngine
+{
+protected:
+  RGWRados* const store;
+  const RGWRemoteAuthApplier::Factory * const apl_factory;
```

Comment:
Cosmetics: it would be nice to have a consistent placing of `*` in pointers declarations.